aws/resource_aws_sns_sms_preferences.go

```diff
@@ -0,0 +1,159 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sns"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsSnsSmsPreferences() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsSnsSmsPreferencesSet,
+		Read:   resourceAwsSnsSmsPreferencesGet,
+		Update: resourceAwsSnsSmsPreferencesSet,
+		Delete: resourceAwsSnsSmsPreferencesDelete,
+
+		Schema: map[string]*schema.Schema{
+			"monthly_spend_limit": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"delivery_status_iam_role_arn": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"delivery_status_success_sampling_rate": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"default_sender_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"default_sms_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"usage_report_s3_bucket": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+const resourceId = "aws_sns_sms_id"
+
+var SMSAttributeMap = map[string]string{
+	"monthly_spend_limit":                   "MonthlySpendLimit",
+	"delivery_status_iam_role_arn":          "DeliveryStatusIAMRole",
+	"delivery_status_success_sampling_rate": "DeliveryStatusSuccessSamplingRate",
+	"default_sender_id":                     "DefaultSenderID",
+	"default_sms_type":                      "DefaultSMSType",
+	"usage_report_s3_bucket":                "UsageReportS3Bucket",
+}
+
+var SMSAttributeDefaultValues = map[string]string{
+	"monthly_spend_limit":                   "",
+	"delivery_status_iam_role_arn":          "",
+	"delivery_status_success_sampling_rate": "",
+	"default_sender_id":                     "",
+	"default_sms_type":                      "",
+	"usage_report_s3_bucket":                "",
+}
+
+func resourceAwsSnsSmsPreferencesSet(d *schema.ResourceData, meta interface{}) error {
+	snsconn := meta.(*AWSClient).snsconn
+
+	log.Printf("[DEBUG] SNS Set SMS preferences")
+
+	monthlySpendLimit := d.Get("monthly_spend_limit").(string)
+	monthlySpendLimitInt, _ := strconv.Atoi(monthlySpendLimit)
+	deliveryStatusIamRoleArn := d.Get("delivery_status_iam_role_arn").(string)
+	deliveryStatusSuccessSamplingRate := d.Get("delivery_status_success_sampling_rate").(string)
+	deliveryStatusSuccessSamplingRateInt, _ := strconv.Atoi(deliveryStatusSuccessSamplingRate)
+	defaultSenderId := d.Get("default_sender_id").(string)
+	defaultSmsType := d.Get("default_sms_type").(string)
+	usageReportS3Bucket := d.Get("usage_report_s3_bucket").(string)
+
+	// Validation
```

Comment:
Please move validation in to `ValidateFunc` in the schema above like this: https://github.com/terraform-providers/terraform-provider-aws/blob/3f9a830038f20f819ba947e90770c50d470d164b/aws/resource_aws_s3_bucket.go#L249

(for these other validations as well, more validations available in the `validation` package, int ranges, string in slice, etc)